packages/wallet/src/TxBuilder/buildTx.ts

```diff
@@ -0,0 +1,171 @@
+import { Cardano, NotImplementedError } from '@cardano-sdk/core';
+import { firstValueFrom } from 'rxjs';
+
+import {
+  DeepPartial,
+  MaybeValidTx,
+  OutputBuilder,
+  OutputValidationMinimumCoin,
+  OutputValidationTokenBundleSize,
+  SignedTx,
+  TxBodyValidationError,
+  TxBuilder,
+  TxOutValidationError
+} from './types';
+import { InitializeTxResult, ObservableWallet } from '../types';
+import { OutputBuilderSaw, toOutputValidationError } from './OutputBuilder';
+import { OutputValidator, StakeKeyStatus, createWalletUtil, deepEquals } from '../services';
+import { SignTransactionOptions } from '../KeyManagement';
+
+/**
+ * Minimal sub-type of ObservableWallet that is used by TxBuilder
+ */
+export type TxBuilderObservableWallet = Pick<
+  ObservableWallet,
+  'utxo' | 'protocolParameters$' | 'finalizeTx' | 'initializeTx' | 'submitTx' | 'delegation'
+>;
+
+interface CreateSignedTxProps {
+  wallet: TxBuilderObservableWallet;
+  initializeTxResult: InitializeTxResult;
+  signOptions?: SignTransactionOptions;
+  auxiliaryData?: Cardano.AuxiliaryData;
+}
+
+const createSignedTx = async ({
+  wallet,
+  initializeTxResult,
+  signOptions,
+  auxiliaryData
+}: CreateSignedTxProps): Promise<SignedTx> => {
+  const tx = await wallet.finalizeTx({ auxiliaryData, props: initializeTxResult, signOptions });
+  return {
+    submit: () => wallet.submitTx(tx),
+    tx
+  };
+};
+
+class TxBuilderSaw implements TxBuilder {
+  partialTxBody: Partial<Cardano.NewTxBodyAlonzo> = {};
+  auxiliaryData?: Cardano.AuxiliaryData;
+  extraWitness?: Partial<Cardano.Witness>;
+
+  #observableWallet: TxBuilderObservableWallet;
+  #util: OutputValidator;
+
+  constructor(observableWallet: TxBuilderObservableWallet, util: OutputValidator) {
+    this.#observableWallet = observableWallet;
+    this.#util = util;
+  }
+
+  addOutput(txOut: Cardano.TxOut): TxBuilder {
+    this.partialTxBody.outputs = [...(this.partialTxBody.outputs || []), txOut];
+    return this;
+  }
+
+  removeOutput(txOut: Cardano.TxOut): TxBuilder {
+    this.partialTxBody.outputs = this.partialTxBody.outputs?.filter((output) => !deepEquals(output, txOut));
+    return this;
+  }
+
+  buildOutput(txOut?: DeepPartial<Cardano.TxOut>): OutputBuilder {
+    return new OutputBuilderSaw(this.#util, txOut);
+  }
+
+  async delegate(poolId: Cardano.PoolId): Promise<TxBuilder> {
+    // REVIEW: should we have an error if no reward account is found. Not sure it's valid...
+    //   Should I be more defensive here and handle an error or empty rewardAccounts?
+    const rewardAccounts = await firstValueFrom(this.#observableWallet.delegation.rewardAccounts$);
+
+    // Discard previous delegation and prepare for new one
+    this.partialTxBody.certificates = [];
+
+    for (const rewardAccount of rewardAccounts) {
+      // REVIEW: Is it ok to create certificates for all the reward accounts?
```

Comment:
Yes, OK for now. Make sure to add a doc comment for this method to highlight this behavior